[linky] Fixes for change in Enedis API on 2024 December 20

lo92fr opened the pull request:
# Fixes for Linky Enedis December web site change

This pull request is about fixing the linky addon after the change that was made on Enedis side to the Web Site.
This is linked to issue:
https://github.com/openhab/openhab-addons/issues/17936

# Description

Mainly, changes are :
    - URL for data is now mes-mesures-prm and not mes-mesures.
    - Dto format have changed : mainly merge data & date, field renaming, and moving.
    - Some changes on date format.

lolodomo commented:
Thank you @lo92fr , I will review and test.

lolodomo commented:
I still have Java 17 and Java 21 on my dev PC so it should be easy to build a JAR for Java 17.
But in case I would have only Java 21, do you know what is the maven command to execute to build a JAR compatible with Java 17 / OH 4.3 ?
@holgerfriedrich I am sure you know ;)

holgerfriedrich commented:
@lolodomo Hopefully it is just:

 `mvn install -Pj17 -Dohc.version=4.3.0`

https://community.openhab.org/t/openhab-5-x-snapshot-requires-java-21/160890/5

lolodomo commented:
@lo92fr : does not work for me => Error requesting 'https://alex.microapplications.enedis.fr/mes-mesures-prm/api/private/v1/personnes/xxxxx/prms/yyyyy/donnees-energetiques?mesuresTypeCode=ENERGIE&mesuresCorrigees=false&typeDonnees=CONS&dateDebut=2023-01-01': {"message":"Internal server error. Please try again","code":"INTERNAL_SERVER_ERROR"} 

Will see if it work in next tries (next hours).

lo92fr commented:
> if it work in next tries (next

It still work on my side, strange.
Does the xxx in URL correctly formatted with something like MPSxxxYYY ?

Laurent.

lolodomo commented:
> Does the xxx in URL correctly formatted with something like MPSxxxYYY ?

3 letters then 3 numbers and then 3 letters.
Starts by N in my case.
If I correctly remember, it is the reference of my counter.

lo92fr commented:
If it's still does not work, can you please:

- check that you have the new version of the website. Look like this:
![image](https://github.com/user-attachments/assets/2fcf8f22-18fb-4131-938c-39b96f6e7025)

- When you load the page, take a look on the browser debugger, new URL appears at the bottom of the list.
Is it the same and does it works correctly in your browser session ? 

![image](https://github.com/user-attachments/assets/33020218-ba93-4c43-a4ba-1e53438df1dd)

 Thks,
     Laurent.

lolodomo commented:
The requests are similar, the binding request has an additional parameter "dateDebut".
When I run in my browser session while I am logged to the website, the binding URL works!

lolodomo commented:
Now the binding is working! Even after I disconnect from the website in my browser.
Maybe you need to connect at least one time to the new website version ?

lolodomo commented on bundles/org.openhab.binding.linky/src/main/java/org/openhab/binding/linky/internal/handler/LinkyHandler.java:
```diff
@@ -211,8 +216,9 @@ private synchronized void updatePowerData() {
-                updatekVAChannel(PEAK_POWER, days.datas.get(days.datas.size() - 1));
-                updateState(PEAK_TIMESTAMP, new DateTimeType(days.periodes.get(days.datas.size() - 1).dateDebut));
+                updatekVAChannel(PEAK_POWER, days.datas.get(days.datas.size() - 1).valeur);
+                updateState(PEAK_TIMESTAMP, new DateTimeType(
+                        days.datas.get(days.datas.size() - 1).dateDebut.atZone(this.timeZoneProvider.getTimeZone())));
```
This was changed very recently, just before releasing 4.3. We don't need to pass a ZonedDateTipe; DateTimeType just needs an Instant as parameter. So you do not need TimeZoneProvider .

lo92fr replied:
Hello @lolodomo,

I'm not sure to understant. Do you say to use the DateTimeType constructor that take an instant, something like this.
 updateState(PEAK_TIMESTAMP,
                        new DateTimeType(days.datas.get(days.datas.size() - 1).dateDebut.toInstant(zoneOffset)));

The matter is that I still need to find a zoneOffset in this case.
The Enedis API return a LocalDateTime at offset of the current zone.
The DateTimeType if I undestand correctly want a universal time.

So somewhere we need to convert from local to universal using a timezone ?

Laurent.

lolodomo replied:
@jlaur is our specialist with date. Please advice.

jlaur replied:
If `days.datas.get(days.datas.size() - 1).dateDebut` is a `LocalDateTime`, the implementation is already correct. If it is an `Instant` there is no need to convert to `ZonedDateTime`.

lolodomo replied:
> the implementation is already correct

@jlaur : you mean @lo92fr proposed code is correct ?
I am a little lost, why should we consider the timezone setup in server ?
Imagine that for a strange reason I setup timezone in OH server to UTC+5. I am convinced Enedis provides dates in France timezone, that is UTC+1.

jlaur replied:
> @jlaur : you mean @lo92fr proposed code is correct ?

Yes, that's what I meant, but purely from a technical perspective, i.e. if you only have a `LocalDateTime`, you also need a `ZoneId` regardless of if you want an `Instant` or `ZonedDateTime`.

> I am a little lost, why should we consider the timezone setup in server  Imagine that for a strange reason I setup timezone in OH server to UTC+5. I am convinced Enedis provides dates in France timezone, that is UTC+1

Indeed, if the local datetime is from a specific server, you need to use the time-zone from that server. This is similar:

https://github.com/openhab/openhab-addons/blob/main/bundles%2Forg.openhab.binding.energidataservice%2Fsrc%2Fmain%2Fjava%2Forg%2Fopenhab%2Fbinding%2Fenergidataservice%2Finternal%2FEnergiDataServiceBindingConstants.java#L88-L90

For a local device, it would make sense to assume same time-zone as openHAB.

lolodomo replied:
@lo92fr : WDYT ? Shouldn't you use the timezone from France rather than the timezone setup in OH server ?

lo92fr replied:
@lolodomo:

At the first, the change was coming from the fact that the new API was returning a datetime without precision of the timezone, something formated like "dateDebut": "2022-01-04T00:00", "dateFin": "2022-01-04T23:59:59.000000999".
The previous API version was returning directly a ZonedDateTime.

So to keep the same call on updateState that before, I convert from the LocalDateTime return by API, assuming that the value return by the API is a LocalDateTime at the zone where the user is located, and that certainly is openhab installation match the same zone.

I don't hard code GMT+1 has the timezone because I assume in some case that the zone can be different.
for exemple, for people leaving in DOM-TOM territory, the timezone will certainly be different.

Laurent.

jlaur replied:
How do they know the time-zone of the user? What happens if you configure your time-zone in openHAB to - 08:00 or something like that?

lo92fr replied:
This can be a problem. I'm not sure of the time-zone of the user and / or the data returned by the API.
But i assume in most case this is aligned because the user as setup the timezone in openhab to the location where he is leaving, that have good chance to be also the location of the linky.

jlaur replied:
Can you configure your home location and/or time-zone for your Linky user account? It should be possible to figure out how they determine the time-zone by changing something for the user account and by setting a different time-zone in openHAB. If none of this affects the time-zone for their local datetimes, I guess time-zone for France can be assumed.

lo92fr replied:
Hello @jlaur,

No there is no way to configure the home location / time-zone in the eneids account.
I think this is a hidden fixed value that is link to your PDL (Point of Delivery) that is linked to your postal address.

So I think the best is to assume that the api just return a local Datetime, and that the openhab server is on the same timezone.

The only scenario I can see that can be bad is someone having two different house : let say one in France, and one in another timezone like Guadeloupe, and that at the same time will use only one openhab instance for the two house.
In this case, result can be wrong.

Laurent.

clinique replied:
Couldn't the timezone be a configuration setting of the PDL thing ? Defaulted to France/Paris ?

lo92fr replied:
Hello @clinique,

Yes, this a solution.
Like this we will be able to handle specific case.

Laurent.

lo92fr replied:
@clinique, @jlaur, 

I've implements the config for timezone in this commit.
Let me know if it's ok for you.
https://github.com/openhab/openhab-addons/pull/17945/commits/4281d5e6014f260454293da01649652d70d9b4b2

Laurent.

lolodomo commented:
The fix fails for me.
It worked just one time apparently because.i previously opened a session on the website but the next days it failed.

lolodomo commented:
While I installed 4.3.1, I updated again the linky bundle and it seems to work well yesterday and also today!
So maybe in my previous test, I did not take care and the previous jar was reloaded...

jlaur commented on bundles/org.openhab.binding.linky/src/main/java/org/openhab/binding/linky/internal/handler/LinkyHandler.java:
```diff
@@ -150,6 +158,19 @@ public void initialize() {
+        Configuration thingConfig = getConfig();
+
+        Object val = thingConfig.get("timezone");
+        if (val == null || "".equals(val)) {
```
```suggestion
        if (val == null || val.isBlank()) {
```

lo92fr replied:
fix on next commit

jlaur commented on bundles/org.openhab.binding.linky/src/main/resources/OH-INF/thing/thing-types.xml:
```diff
@@ -34,6 +34,16 @@
+			<parameter name="timezone" type="text" required="false">
+				<label>timezone</label>
+				<description>The timezone associated with your Point of delivery.
+					Will default to openhab default timezone.
```
```suggestion
					Will default to openHAB default timezone.
```

lo92fr replied:
fix on next commit

jlaur commented on bundles/org.openhab.binding.linky/src/main/resources/OH-INF/thing/thing-types.xml:
```diff
@@ -34,6 +34,16 @@
+				<description>The timezone associated with your Point of delivery.
+					Will default to openhab default timezone.
+					You will
+					need to change this if your linky is located in a different timezone that your openhab location.
```
Not sure if Linky should start with capital letter, but most likely?
```suggestion
					need to change this if your Linky is located in a different timezone that your openHAB location.
```

lo92fr replied:
fix on next commit

lo92fr replied:
I've also regenerated the i18n files

jlaur reviewed (commented):
Minor comments related to the time-zone commit. The I18 properties file should also be [regenerated](https://v40.openhab.org/docs/developer/utils/i18n.html#generating-i18n-properties-file).

lo92fr commented:
Hello @jlaur, Hello @lolodomo,

Any news on this ?

Thanks,
     Laurent. @

jlaur commented:
> Any news on this ?

My comments are resolved, but I have not reviewed this PR, so @lolodomo will be needed for finalizing the review and merging the PR.

lolodomo commented:
The major problem is that unfortunately this fix does not work well !
It can work one day after a fresh OH install and then does not work anymore. This is probably a problem of cookie or similar.
I remember that we had exactly the same issue with the previous fix ... until @lo92fr fixed something more.

Am I alone with this PR not working more than one day ?

lo92fr commented:
> The major problem is that unfortunately this fix does not work well ! It can work one day after a fresh OH install and then does not work anymore. This is probably a problem of cookie or similar. I remember that we had exactly the same issue with the previous fix ... until @lo92fr fixed something more.
> 
> Am I alone with this PR not working more than one day ?

Have to said that I'm not totally sure myself if the fix is ok or not.
I'm working on my production server with the linkyv2 branch, and have disconnection every 24 hours.
I was thinking it was because of the workflow change that i've maded on linkyv2 branch (linky v2 keep the connection opened, but main branch disconnect / reconnect at every 24 hours refresh), but finally perhaps not.

@lolodomo : what error do you have after one day, an http 500 Internal Error ?
Anything in the log ?

What I try so far if to make the refresh more frequent, I see that if I change the refresh frequence from 24h to 3h, I can trigger the error as weel after only 3 hours.

I've also try in this case to clear all cookie, and redo the connectionInit phase, but this don't seem sufficient to fix the error.
But in v2, when I hit the error, If i stop the thing, and restart it, the connection restart to work correctly.

So I thing there is something that expire after some session time, and stay in memory, and that is not reinitialized correctly.

I will continue to narrow done this issue in the next few days, and will tell you if I find something.

Laurent.

lolodomo commented:
> what error do you have after one day, an http 500 Internal Error ?

{"message":"Internal server error. Please try again","code":"INTERNAL_SERVER_ERROR"}

lo92fr commented:
> > what error do you have after one day, an http 500 Internal Error ?
> 
> {"message":"Internal server error. Please try again","code":"INTERNAL_SERVER_ERROR"}

Hello all, @lolodomo 

Perhaps find a fix on this new error.
It seems that we need to call the getPrmInfo each time we update the data.
Today, we only do it when initializing the handler.

My assomption is that the new website implementation keep some state data, and invalidate them after a few hours.
This state data is certainly initialize when we call this getPrmInfo endpoint.

I've doned some test on v2 branch, seems to fix the problems on my side.
Will need to be tested also on this fix branch to be sure.

Laurent.

lolodomo commented:
@lo92fr : can you please rebase your branch so that we can compile it again ?

lo92fr commented:
> do it when initializing the handler.
> 
> My assomption

Helo @lolodomo,

Done the rebasing just right now.
But note that my fix from last week is not sufficient to totally fix our issues.
I've done some more modifications on the linkyv2 branch that seems to be promising.
I will try to backport them to fix branch this afternoon, or somewhere laster this week if I don't find time today.

Laurent.

lolodomo commented:
I just build and deploy the last jar and yes you're right, it still does not work, always the same internal server error.

lo92fr commented:
> does not work, always the same internal server error.

Hello @lolodomo,

Just to be sure to understand : you have the 500 Internal Server errors just after starting the binding, or on the nightly refresh at 1'oclock ?
If it's just after the start, can you send me your log, with serverity config to DEBUG ?

On my side, it's working since several day without error, refreshing data correctly every day.
But it's on v2 branch, so I switch this morning my server on the fix branch to check it better.

Also, I have to done some modifications on the ExperingDayCache.
The reason is that from my experiments, enedis site respond at 1'oclock without error, but the data is incomplete.
For exemple, I've got NULL value for day-1 consumption.

So I've made some modification on the logic of the Cache : if I detect that data is incomplete, I retry to read data after the REFRESH_INTERVAL_IN_MIN (120mn). So it will try at 1'oclock, 3'oclock, 5'oclock until he get the correct data.

I've also modifify the nextExpireAt logic a little, the idea is not to do all the request at the exact same hours for all users. But to randomize them a little to distribute the request around the central hour.

This change is on my last commit in the fix branch. I've backport them this morning from v2.

Laurent.

lo92fr commented:
> > does not work, always the same internal server error.
> 
> Hello @lolodomo,
> 
> Just to be sure to understand : you have the 500 Internal Server errors just after starting the binding, or on the nightly refresh at 1'oclock ? If it's just after the start, can you send me your log, with serverity config to DEBUG ?
> 
> On my side, it's working since several day without error, refreshing data correctly every day. But it's on v2 branch, so I switch this morning my server on the fix branch to check it better.
> 
> Also, I have to done some modifications on the ExperingDayCache. The reason is that from my experiments, enedis site respond at 1'oclock without error, but the data is incomplete. For exemple, I've got NULL value for day-1 consumption.
> 
> So I've made some modification on the logic of the Cache : if I detect that data is incomplete, I retry to read data after the REFRESH_INTERVAL_IN_MIN (120mn). So it will try at 1'oclock, 3'oclock, 5'oclock until he get the correct data.
> 
> I've also modifify the nextExpireAt logic a little, the idea is not to do all the request at the exact same hours for all users. But to randomize them a little to distribute the request around the central hour.
> 
> This change is on my last commit in the fix branch. I've backport them this morning from v2.
> 
> Laurent.

Hello @lolodomo, @jlaur, @clinique,

Just a little news about the fix.

I've reverted some of my last change on ExperingDayCache as I realized that the missingData  logic I add was duplicate with check in getConsumptionAfterChecks.

I've also modified condition in getMetaData so we always call the getPrmInfos / getPrmDetails even if we already have the information. With this last fix, it solve the 500 nightly error on my side.

I've deploy the branch on my production openhab, and it now working well since 2 day without 500 error, and getting fresh value every night :)

I will keep it until monday to be sure everything is okl, but I think you perhaps can get another try to this new version.

Laurent.

lolodomo commented:
Thank you Laurent, I just deployed the new version and the binding is ONLINE. I will see if it survives at least 3 days. I am starting a new review of the PR.

lolodomo commented on bundles/org.openhab.binding.linky/src/main/java/org/openhab/binding/linky/internal/dto/ConsumptionReport.java:
```diff
@@ -22,43 +23,41 @@
  * returned by API calls
  *
  * @author Gaël L'hopital - Initial contribution
+ * @author Laurent ARNAL - fix to handle new Dto format after enedis site modifications
```
Rather Arnal

lo92fr replied:
done on next commit

lolodomo commented on bundles/org.openhab.binding.linky/src/main/java/org/openhab/binding/linky/internal/handler/LinkyHandler.java:
```diff
@@ -65,7 +69,12 @@
 @NonNullByDefault
 public class LinkyHandler extends BaseThingHandler {
-    private static final int REFRESH_FIRST_HOUR_OF_DAY = 1;
+    private final TimeZoneProvider timeZoneProvider;
+    private ZoneId zoneId = ZoneId.systemDefault();
+
+    private static final Random randomNumbers = new Random();
+    private static final int REFRESH_HOUR_OF_DAY = 1;
+    private static final int REFRESH_MINUTE_OF_DAY = randomNumbers.nextInt(60);
```
Minor but the recommendation is to have all the static first, then the non static final.

lo92fr replied:
done on next commit

lolodomo commented on bundles/org.openhab.binding.linky/src/main/java/org/openhab/binding/linky/internal/handler/LinkyHandler.java:
```diff
@@ -107,12 +119,15 @@ public LinkyHandler(Thing thing, LocaleProvider localeProvider, Gson gson, HttpC
+        this.cachedPowerData = new ExpiringDayCache<>("power cache", REFRESH_HOUR_OF_DAY, REFRESH_MINUTE_OF_DAY, () -> {
+            // We request data for yesterday and the day before yesterday, even if the data for the day before
+            // yesterday
             // is not needed by the binding. This is only a workaround to an API bug that will return
-            // INTERNAL_SERVER_ERROR rather than the expected data with a NaN value when the data for yesterday is not
+            // INTERNAL_SERVER_ERROR rather than the expected data with a NaN value when the data for yesterday
+            // is not
             // yet available.
-            // By requesting two days, the API is not failing and you get the expected NaN value for yesterday when the
+            // By requesting two days, the API is not failing and you get the expected NaN value for yesterday
+            // when the
             // data is not yet available.
```
If you cut comments with a smaller size, please try to group certain lines

lolodomo commented on bundles/org.openhab.binding.linky/src/main/java/org/openhab/binding/linky/internal/handler/LinkyHandler.java:
```diff
@@ -124,32 +139,49 @@ public LinkyHandler(Thing thing, LocaleProvider localeProvider, Gson gson, HttpC
         logger.debug("Initializing Linky handler.");
         updateStatus(ThingStatus.UNKNOWN);
 
+        // update the timezone if not set to default to openhab default timezone
+        Configuration thingConfig = getConfig();
+
+        String val = (String) thingConfig.get("timezone");
```
Look at just lines just after, a better way to get the timezone value is:
```
getConfigAs(LinkyConfiguration.class).timezone
```

lo92fr replied:
I do it this way because of the updating on line 180 (updateConfiguration(thingConfig)
If I remember, if I use the other way getConfigAs(), I'm not able to update the config after it

lo92fr replied:
Ok finally rewrite it to remove the update logic.

lolodomo commented on bundles/org.openhab.binding.linky/src/main/java/org/openhab/binding/linky/internal/handler/LinkyHandler.java:
```diff
@@ -124,32 +139,49 @@ public LinkyHandler(Thing thing, LocaleProvider localeProvider, Gson gson, HttpC
+        Configuration thingConfig = getConfig();
+
+        String val = (String) thingConfig.get("timezone");
+        if (val == null || val.isBlank()) {
+            zoneId = this.timeZoneProvider.getTimeZone();
+            thingConfig.put("timezone", zoneId.getId());
+        } else {
+            zoneId = ZoneId.of(val);
+        }
+
+        updateConfiguration(thingConfig);
```
I do not like this useless config update, in particular for things defined in read-only config files.
I believe you can just keep the computed zone in your variable zoneId without updating the thing configuration.

lo92fr replied:
Ok finally rewrite it to remove the update logic.

lolodomo commented on bundles/org.openhab.binding.linky/src/main/java/org/openhab/binding/linky/internal/handler/LinkyHandler.java:
```diff
@@ -193,26 +212,48 @@ public void initialize() {
+                disconnect();
+            }
+        } catch (LinkyException e) {
+            logger.error("Exception occurs during data update {}", e.getMessage(), e);
```
Please update to DEBUG level.

lo92fr replied:
done on next commit

lolodomo commented on bundles/org.openhab.binding.linky/src/main/java/org/openhab/binding/linky/internal/handler/LinkyHandler.java:
```diff
@@ -426,26 +466,32 @@ public synchronized void handleCommand(ChannelUID channelUID, Command command) {
+                        break;
+                }
+            } catch (LinkyException ex) {
+                logger.error("Unable to handleCommand refresh", ex);
```
Please update to DEBUG level.

lo92fr replied:
done on next commit

lolodomo commented on bundles/org.openhab.binding.linky/src/main/java/org/openhab/binding/linky/internal/handler/LinkyHandler.java:
```diff
@@ -193,26 +212,48 @@ public void initialize() {
+            UserInfo userInfo = api.getUserInfo();
+            PrmInfo prmInfo = api.getPrmInfo(userInfo.userProperties.internId);
+            PrmDetail details = api.getPrmDetails(userInfo.userProperties.internId, prmInfo.idPrm);
+            updateProperties(Map.of(USER_ID, userInfo.userProperties.internId, PUISSANCE,
+                    details.situationContractuelleDtos[0].structureTarifaire().puissanceSouscrite().valeur() + " kVA",
+                    PRM_ID, prmInfo.idPrm));
```
Thing properties will now be updated more often than before.
This is probably not a big problem as the binding is not updating data very often.

lo92fr replied:
Yes, I know.
But it seems the only way to have this 500 error fixed, and as the refresh rate is only 2 hours, it's not really a big problem.

lolodomo commented on bundles/org.openhab.binding.linky/src/main/java/org/openhab/binding/linky/internal/handler/LinkyHandler.java:
```diff
@@ -474,50 +520,36 @@ public synchronized void handleCommand(ChannelUID channelUID, Command command) {
-            throw new LinkyException("Invalid consumptions data: not any data for each day period");
-        }
-        if (consumption.aggregats.weeks.periodes.isEmpty()) {
+        if (consumption.aggregats.weeks != null && consumption.aggregats.weeks.datas.isEmpty()) {
```
Shouldn't it be rather ?
```
if (consumption.aggregats.weeks == null || consumption.aggregats.weeks.datas.isEmpty()) {
```

lolodomo replied:
In case I am right, the same change should be applied to next tests in this method.

lo92fr replied:
I don't think. This is not really revelent on this branch, because this structure days / weeks / months / years is always different from null.

But on v2 branch, we introduce also an hours agregate, and base on api calls, we can have an agregate with:

hour!=null
days, week, months, years = null

or

hour=null
days, week, months, years != null

If we look at main branch, the test != null was not there at all. We was only testing .isEmpty()

What we want to be sure by this condition is that if the API return something for the agregat (!=null), this agregat is not empty, that will be an anormal situation.

So I propose to keep it like this.

lolodomo commented on bundles/org.openhab.binding.linky/src/main/java/org/openhab/binding/linky/internal/handler/LinkyHandler.java:
```diff
@@ -536,12 +568,17 @@ private void logData(Aggregate aggregate, String title, boolean withDateFin, Dat
+                logger.debug("{} {} value {}", title, aggregate.datas.get(index).dateDebut.format(dateTimeFormatter),
+                        aggregate.datas.get(index).valeur);
+            }
+        } catch (Exception e) {
+            logger.error("error during logData", e);
```
What kind of exception are you trying to catch ?

lo92fr replied:
It was NullPointerException.
At some place, I it this one durring my test.
But I don't thing we really need it, so I propose to remove this try/catch that was not there in main branch.

lolodomo commented on bundles/org.openhab.binding.linky/src/main/resources/OH-INF/i18n/linky.properties:
```diff
@@ -41,6 +43,6 @@ channel-type.linky.power.label = Yesterday Peak Power
 channel-type.linky.timestamp.label = Timestamp
 
-# Thing status descriptions
+# thing status descriptions
```
If it is, that's fine.

lo92fr replied:
Yes, I think, I don't remember to have change this manually.
But at some place, I've triggered the generation script.

lo92fr commented on bundles/org.openhab.binding.linky/src/main/resources/OH-INF/thing/thing-types.xml:
```diff
@@ -34,6 +34,16 @@
 				<description>Authentication ID delivered after the captcha (see documentation).</description>
 			</parameter>
+			<parameter name="timezone" type="text" required="false">
+				<label>timezone</label>
```
fix on next commit

lo92fr commented on bundles/org.openhab.binding.linky/src/main/resources/OH-INF/thing/thing-types.xml:
```diff
@@ -34,6 +34,16 @@
+			<parameter name="timezone" type="text" required="false">
+				<label>timezone</label>
+				<description>The timezone associated with your Point of delivery.
+					Will default to openHAB default timezone.
+					You will
+					need to change this if your Linky is located in a different timezone that your openHAB location.
+					You can use an
+					offset, or a label like Europe/Paris</description>
```
Yes, fix in next commit

lo92fr commented on bundles/org.openhab.binding.linky/src/main/resources/OH-INF/thing/thing-types.xml:
```diff
@@ -34,6 +34,16 @@
 				<label>Auth ID</label>
 				<description>Authentication ID delivered after the captcha (see documentation).</description>
 			</parameter>
+			<parameter name="timezone" type="text" required="false">
```
done on next commit.

lo92fr commented on bundles/org.openhab.binding.linky/src/main/java/org/openhab/binding/linky/internal/LinkyHandlerFactory.java:
```diff
@@ -55,21 +61,42 @@
+                        try {
+                            return LocalDateTime.parse(json.getAsJsonPrimitive().getAsString(),
+                                    LINKY_LOCALDATETIME_FORMATTER);
+                        } catch (Exception ex) {
```
fix on next commit

lo92fr commented on bundles/org.openhab.binding.linky/src/main/java/org/openhab/binding/linky/internal/api/EnedisHttpApi.java:
```diff
@@ -261,10 +294,15 @@ private <T> T getData(String url, Class<T> clazz) throws LinkyException {
         } catch (JsonSyntaxException e) {
             logger.debug("Invalid JSON response not matching {}: {}", clazz.getName(), data);
             throw new LinkyException(e, "Requesting '%s' returned an invalid JSON response", url);
+        } catch (Exception e) {
+            logger.error("Error {}: {}", clazz.getName(), data, e);
+            throw new LinkyException(e, "Requesting '%s' returned an invalid JSON response", url);
```
Yes, I don't see good reason to have this second catch block.
So I removed it.

lo92fr commented:
Hello @lolodomo,

I've fixed the issue of your last review.

Laurent.

lolodomo commented:
@lo92fr : please rerun the i18n tool to take into consideration your change of label.
Then that will be fine for me.

The binding was still "alive" this morning.
I plan to wait for one another day before merging.

lo92fr commented:
> @lo92fr : please rerun the i18n tool to take into consideration your change of label. Then that will be fine for me.
> 
> The binding was still "alive" this morning. I plan to wait for one another day before merging.

Hello @lolodomo,

This is done for regeneration of i18n.

Laurent.
